packages/agent/src/oidc-v2.ts

```diff
@@ -0,0 +1,689 @@
+ */
+function buildOidcUrl({
+  baseURL,
+  endpoint,
+  authParam,
+  tokenParam,
+}: {
+  baseURL: string;
+  endpoint: OidcEndpoint;
+  authParam?: string;
+  tokenParam?: string;
+}) {
+  switch (endpoint) {
+    /** 1. client sends {@link PushedAuthRequest} & receives {@link PushedAuthResponse} */
+    case 'pushedAuthorizationRequest':
+      return appendPathToUrl({
+        path : 'par',
+        url  : baseURL,
+      });
+    /** 2. provider gets {@link HybridAuthRequest} */
+    case 'authorize':
+      return authParam
+        ? appendPathToUrl({
+          path : `authorize/${authParam}.jwt`,
+          url  : baseURL,
+        })
+        : '';
+    /** 3. provider sends {@link HybridAuthResponse} */
+    case 'callback':
+      return appendPathToUrl({
+        path : `callback`,
+        url  : baseURL,
+      });
+    /**  4. client gets {@link HybridAuthResponse */
+    case 'token':
+      return tokenParam
+        ? appendPathToUrl({
+          path : `token/${tokenParam}.jwt`,
+          url  : baseURL,
+        })
+        : '';
+    // TODO: metadata endpoints?
+    default:
+      return '';
+  }
+}
+
+/**
+ * Generates a random string value used to associate a Client authorization
+ * request with the Identity Provider's response callback.
+ *
+ * @returns A random state as a Base64Url encoded string.
+ */
+function generateRandomState() {
+  const randomStateu8a = utils.randomBytes(12);
+  const ramdomStateb64url = Convert.uint8Array(randomStateu8a).toBase64Url();
+
+  return { randomStateu8a, ramdomStateb64url };
+}
+
+/**
+ * Calculates a value based on the given input that can be used to associate a
+ * Client session with an ID Token, and to mitigate replay attacks. Per the
+ * OpenID Connect Core 1.0 specification, the nonce parameter value should
+ * include per-session state and be unguessable to attackers.
+ *
+ * @see {@link https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes | OpenID Connect Core 1.0, Nonce Implementation Notes}
+ * @returns A derived nonce as a Base64Url encoded string.
+ */
+async function deriveNonceFromInput(state: Uint8Array) {
+  const nonceBytes = await Hkdf.deriveKeyBytes({
+    hash         : 'SHA-256',
+    baseKeyBytes : state,
+    length       : 16,
+    salt         : utils.randomBytes(12),
+  });
+
+  const nonce = Convert.uint8Array(nonceBytes).toBase64Url();
+
+  return nonce;
+}
+
+/**
+ * Generates a cryptographically random key called a "code verifier" in
+ * accordance with the RFC 7636 PKCE specification. A unique code verifier
+ * should be created for every authorization request.
+ *
+ * @see {@link https://datatracker.ietf.org/doc/html/rfc7636#section-4.1 | RFC 7636, Client Creates a Code Verifier}
+ *
+ * @returns A random code verifier as a Base64Url encoded string.
+ */
+function generateRandomCodeVerifier() {
+  const codeVerifieru8a = utils.randomBytes(32);
+
+  return { codeVerifieru8a };
```

Comment:
Normally I'd agree with what you're saying but because this is a specification it takes a lot more effort and research to find a spec based number like 32 than it does to just use a bespoke method that enshrines the spec into the code so I'd rather keep this. No problem on the name change

Reply:
done